src/neovim/cursor.ts

```diff
@@ -157,17 +157,51 @@ export default class NeovimCursor {
         this.blink_timer.reset();
     }
 
+    updateCursorSize() {
+        const info = this.store.modeInfo[this.store.mode];
+        if (!info) {
+            return;
+        }
+        const {cursor_shape, cell_percentage} = info;
+        if (!cursor_shape) {
+            return;
+        }
+        const {width, height} = this.store.font_attr;
+        switch (cursor_shape) {
+            case 'horizontal': {
+                const top = height * (1 - cell_percentage / 100);
+                const right = width;
+                const bottom = height;
+                const left = 0;
+                this.element.style.clip = `rect(${top}px ${right}px ${bottom}px ${left}px)`;
+                break;
+            }
+            case 'vertical': {
+                const top = 0;
+                const right = width * (cell_percentage / 100);
+                const bottom = height;
+                const left = 0;
+                this.element.style.clip = `rect(${top}px ${right}px ${bottom}px ${left}px)`;
+                break;
+            }
+            default: {
+                this.element.style.clip = 'auto';
+                break;
+            }
+        }
+    }
+
     private redrawImpl() {
         this.delay_timer = null;
-        const cursor_width = this.store.mode.endsWith('insert') ? (window.devicePixelRatio || 1) : this.store.font_attr.draw_width;
-        const cursor_height = this.store.font_attr.draw_height;
-        const x = this.store.cursor.col * this.store.font_attr.draw_width;
-        const y = this.store.cursor.line * this.store.font_attr.draw_height;
-        const captured = this.screen_ctx.getImageData(x, y, cursor_width, cursor_height);
+        const {draw_width, draw_height} = this.store.font_attr;
+        const x = this.store.cursor.col * draw_width;
+        const y = this.store.cursor.line * draw_height;
+        const captured = this.screen_ctx.getImageData(x, y, draw_width, draw_height);
```

Comment:
Why do you think it is necessary to care about it? This method has nothing to do with the cursor shape any longer.